src/core/lib/support/reference_counted.h

```diff
@@ -0,0 +1,59 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_LIB_SUPPORT_REFERENCE_COUNTED_H
+#define GRPC_CORE_LIB_SUPPORT_REFERENCE_COUNTED_H
+
+#include <grpc/support/sync.h>
+
+#include "src/core/lib/debug/trace.h"
+#include "src/core/lib/support/debug_location.h"
+
+namespace grpc_core {
+
+class ReferenceCounted {
```

Comment:
Suggest renaming this to ReferenceCountedWithTracing, and introducing a lower level ReferenceCounted type that doesn't carry the additional TraceFlag.

My expectation is that as time goes by, we'll move more and more code from ReferenceCountedWithTracing to ReferenceCounted (as we transition away from manual reference counting)

Reply:
I tried this, but I ran into two ugly problems:

1. In order for the `ReferenceCountedWithTracing` class to have access to the current refcount for logging purposes, I would need to either (a) make it a friend of the parent class or (b) provide a protected method to access the current refcount.  The latter is a bit cleaner than the former, but it would mean that we're exposing implementation details into all subclasses.

2. Having methods called `Ref()` and `Unref()` in the `ReferenceCountedWithTracing` class would mask the methods of the same names but with different signatures in the base class.  This would require renaming the methods in the `ReferenceCountedWithTracing` class to something like `RefWithTracing()` and `UnrefWithTracing`, which seems sub-optimal.

As an alternative, I suppose I could simply have two completely unrelated classes with the same interface and functionality, but I'm not wildly enthusiastic about it.  Any suggestions on a better alternative?

Reply:
I think two unrelated classes is fine here. We can parameterize tests for each of them to ensure consistent behavior, and the smart pointer code is common, and arguably more important.

Reply:
Okay, done.